### PR TITLE
Add Florence2 model support and tests

### PR DESCRIPTION
**Pull request by fjaijf**
## Summary
- Add Florence2 model, config, processor, image processor, auto mappings, and tests.
- Add Florence2 SFT-VL config/template wiring.
- Recreate this PR from the fjaijf fork; supersedes #4612 from the previous fork.

## Tests
- Not run in this submission turn; this is a PR recreation from the existing Florence-2-base branch.

### CODE REVIEW

**Comment by fjaijf** [comment]
Closing this PR to recreate it with a clean description from the fjaijf fork.

File: paddleformers/transformers/florence2/modeling.py [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+    def _split_sft_inputs(self, input_ids, labels, attention_mask):
+        source_rows, label_rows = [], []
+        max_source = 1
+        max_target = 1
+        for row, label_row in zip(input_ids.tolist(), labels.tolist()):
+            target_start = next((index for index, value in enumerate(label_row) if value != -100), len(row))
+            # PaddleFormers SFT labels are shifted left once, so the first
+            # supervised label predicts the token after this source position.
+            source = row[: target_start + 1] or [self.config.bos_token_id]
+            target = [value for value in label_row[target_start:] if value != -100]
```
Comment (risemeup1111) [inline paddleformers/transformers/florence2/modeling.py:908]: P1 多轮 SFT 会在这里被静默改写：代码只按第一个非 `-100` 位置截取 source，随后把所有非 `-100` label 拼成一个 decoder target。例如 `input_ids=[10,11,20,21,12,13,30,31,2]`、`labels=[-100,20,21,-100,-100,30,31,2,-100]` 会得到 source `[10,11]`、target `[20,21,30,31,2]`，第二轮用户上下文 `[12,13]` 被丢弃。当前测试只覆盖单轮。请按每个 assistant span 保留相应上下文（或明确拒绝多轮样本），并增加多轮回归测试。

File: paddleformers/transformers/florence2/processor.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        return_tensors="pd",
+        padding=False,
+        truncation=None,
+        max_length=None,
+        **kwargs,
+    ):
+        if images is None:
+            raise ValueError("`images` must be provided to Florence2Processor.")
+        texts = text if isinstance(text, list) else [text or ""]
+        if isinstance(images, list) and len(images) < len(texts):
```
Comment (risemeup1111) [inline paddleformers/transformers/florence2/processor.py:117]: P1 这里既只检查 Python `list`，又只检查 `len(images) < len(texts)`，因此 `text=None` 配两张图、一个 prompt 配两张图，或 batched NumPy/Paddle 图像都会返回文本 batch=1、像素 batch=2 的 `BatchFeature`，随后 `_merge_image_features` 在 batch 维拼接时失败。请先统一规范化图像输入，并确保两侧 batch 相等；`text=None` 可按图像数生成空 prompt，其他不匹配应显式拒绝或按明确规则广播，同时补充 `__call__` 的批量测试。

File: paddleformers/transformers/auto/modeling.py [hunk 3]
```diff
@@ -86,6 +86,7 @@
         ("Gemma3", "gemma3"),
         ("Gemma3Text", "gemma3_text"),
         ("Gemma4Moe", "gemma4_moe"),
+        ("Florence2", "florence2"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/modeling.py:89]: P1 这项通用映射会让默认 `AutoModel.from_pretrained()` 推导出不存在的 `Florence2Model`。官方配置的 `architectures` 是 `Florence2ForConditionalGeneration`，但 `_get_model_class_from_config` 命中 `Florence2` 后固定拼接 `Model`，最终在 `florence2.modeling` 中查找 `Florence2Model` 并抛出 `AttributeError`；当前模块只提供 `Florence2ForConditionalGeneration`。请为 `Model` 任务映射到实际类（或不要注册不可用的通用入口），并增加 `AutoModel` 从真实配置/本地目录加载的测试。

File: paddleformers/transformers/florence2/processor.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                    labels.append(phrase.strip())
+            return {task: {"polygons": polygons, "labels": labels}}
+        if task == "<REGION_PROPOSAL>":
+            values = [int(value) for value in re.findall(r"<loc_(\d+)>", clean_text)]
+            bboxes = [
+                self._dequantize(values[index : index + 4], image_size) for index in range(0, len(values) - 3, 4)
+            ]
+            return {task: {"bboxes": bboxes, "labels": [""] * len(bboxes)}}
+
+        phrase_pattern = r"([^<]+)((?:<loc_\d+>){4,})"
```
Comment (risemeup1111) [inline paddleformers/transformers/florence2/processor.py:208]: P1 `<OPEN_VOCABULARY_DETECTION>` 的合法输出既可能是 bbox，也可能是 polygon；例如 `cat<poly><loc_0>...<loc_999></poly>`。当前 fallback 要求类别后直接连续出现 `<loc_*>`，遇到 `<poly>` 会完全匹配不到并返回空结果。请为该任务同时解析 bbox/polygon（与 Florence-2 的 `description_with_bboxes_or_polygons` 契约一致），并分别覆盖两种生成格式。

File: examples/config/sft-vl/florence2_full_300_steps.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+eval_dataset_path: ./florence2_train.jsonl
+eval_dataset_prob: "1.0"
+max_seq_len: 1024
+packing: false
+mix_strategy: concat
+template_backend: custom
+template: florence2
+
+### model
+model_name_or_path: /home/housaijie/code/Florence-2-base
```
Comment (risemeup1111) [inline examples/config/sft-vl/florence2_full_300_steps.yaml:18]: P2 该示例固定到作者机器的 `/home/housaijie/...`，其他用户直接运行必然找不到模型。这里已启用 `convert_from_hf: true`，请改为可公开解析的模型 ID（如 `microsoft/Florence-2-base`）或仓库内可配置的相对路径。

File: paddleformers/transformers/florence2/modeling.py [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+        loss = None
+        if labels is not None:
+            loss = F.cross_entropy(
+                logits.reshape([-1, self.config.vocab_size]),
+                labels.reshape([-1]),
+                ignore_index=-100,
+            )
+        if not return_dict:
+            return ((loss, logits) if loss is not None else (logits,)) + outputs[1:]
+        return Seq2SeqLMOutput(
```
Comment (risemeup1111) [inline paddleformers/transformers/florence2/modeling.py:746]: P2 当前生成循环不会保存这里返回的 cache：`GenerationMixin.update_model_kwargs_for_generation()` 只从 tuple 或 `CausalLMOutputWithPast` 读取 `past_key_values`，而本实现返回 `Seq2SeqLMOutput`，且 `prepare_inputs_for_generation` 固定 `return_dict=True`。所以后续每一步收到的 cache 都是 `None`，会重复计算完整 decoder 前缀。请补充该输出类型的 cache 传递（或在模型侧覆写 updater），并测试第二步仅输入一个 token 且携带非空 cache。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [多轮 SFT 上下文](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756580687) | P1 | 🚧 |
| 2 | [Processor 批次一致性](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756582541) | P1 | 🚧 |
| 3 | [AutoModel 类映射](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756584258) | P1 | 🚧 |
| 4 | [开放词汇 polygon 解析](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756586022) | P1 | 🚧 |
| 5 | [示例模型路径](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756588964) | P2 | 🚧 |
| 6 | [生成 cache 传递](https://github.com/PaddlePaddle/PaddleFormers/pull/4868#discussion_r3756591814) | P2 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
